### PR TITLE
[Release/3.0] Add harvest validation logic to the all configurations tests

### LINKED ISSUE #39537: [3.0 servicing] Determine what we do to improve package harvesting story for 3.0
See https://github.com/dotnet/arcade/issues/2929

We need to decide how much of this we tool for 3.0 servicing.

### PR DESCRIPTION
**Pull request by joperezr**
Fixes #39537

Master PR with same changes: https://github.com/dotnet/corefx/pull/40717

This will add harvesting validation that will make sure that once we enter into servicing mode, in the case we want to service a specific package which contains an asset being harvested from a different branch, we will ensure that we are harvesting the latest asset. We have recently found several bugs because of this lack of validation, so adding this will ensure that this won't happen again.

cc: @ericstj @ViktorHofer 

### Description
Here is an example to better explain the problem. In Release/2.1 branch, we were building the System.Net.Http.WinHttpHandler package, but we turned off building some of its configurations in order to simplify our build. For these configurations that we turned off, we were instead harvesting them from the latest stable package built out from the release/1.1 branch, which would still live built these configurations. The problem is, if release/1.1 ever had another bug fix in one of those configurations and ship a new package, our release/2.1 build had no way of figuring this out, so new packages from release/2.1 branch would still harvest the assets from the old release/1.1 package that had the bug.

With the validation that I'm adding here, as part of the all configuration tests, we will ensure that if a package is harvesting assets, it will do so from the latest package version from that package era, so that if one of the harvested assets gets serviced, then we will immediately detect it in these tests and fail the build until we bump the version we are harvesting in order to get the fixes as well. This validation proves very useful specially when servicing packages out of different branches, since you have protection to make sure that you are always servicing with latest fixes from all branches.

### Customer Impact:
Changes are infrastructure related so shouldn't have direct impact to the customer. After this goes in, customer will always get latest fixes across different branches on all new package versions.

### Regression? 
No. This is just adding validation to protect against regressions.

### Risk
Low. Most of this change only contains logic to run a test for validation against a regression. The low risk part of this change, is that we are now calculating automatically which package version gets harvested instead of the previous way when we had it checked in directly in the repo.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@joperezr any reason why you don't want the PR to be merged yet? Tell mode for infra doesn't require the template (even though it's helpful in this case) or approval but just a mail being sent out to tactics.

**Comment by joperezr** [comment]
@ViktorHofer I ask @danmosemsft and we were not 100% sure of the current state of tell mode which is why I decided to wait for tactics to give the green light

**Comment by ericstj** [comment]
Can we include the PackageIndex update as well so that we have a good baseline that is consistent with what's shipped in the past?

**Comment by ViktorHofer** [comment]
I believe that's already part of the PR - if you mean the removal of the hard-coded list.

**Comment by joperezr** [comment]
No, he means updating the stable version list which is out of date, I'll will calculate that diff and let you guys know how big it is.

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json [hunk 1]
```diff
@@ -3458,7 +3532,6 @@
     },
     "System.Reflection.Emit.Lightweight": {
       "StableVersions": [
-        "4.0.0",
```
Comment (ericstj) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: this delete is causing the CI failure.

Comment (ericstj) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: Hmm, looks like these were unlisted.  I think that was due to a misdirection around how to solve ref.emit.  Folks unlisted originally, but then backed up and decided to ship a new version with not-supported assembly in 5.0.  I wonder if we should relist those...

Comment (joperezr) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: I see, for the time being I re added this stable version back. We can decide whether or not to re-list the package in nuget separably.

**Comment by ViktorHofer** [comment]
> for tactics to give the green light

ready? 🎉

**Comment by joperezr** [comment]
@ViktorHofer yup 😄

**Pull request merged**